Add 16a4w_block QAT config

sxu opened the pull request:
Summary: Introduce a FakeQuantizer subclass. It falls back to LPBQ observer's `convert`. `_derived_bias_quant_spec` also looks for it to correctly derive bias scale.

Open to suggestions on how to test. Naveen launched a QAT run and it seems to produce reasonable results.

Differential Revision: D87194388

billmguo commented:
@winskuo-quic  can you review and approve this diff?

haowhsu-quic commented on backends/qualcomm/quantizer/observers/per_block_param_observer.py:
```diff
@@ -89,3 +89,62 @@ def calculate_qparams(self) -> Tuple[torch.Tensor, torch.Tensor]:
+        self.eps = eps
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        if x.numel() == 0:
```
Would it be simpler if calling `torchao.quantization.quant_primitives._fake_quantize_affine` directly?

sxu replied:
Updated to use _fake_quantize_affine.

DannyYuyang-quic commented:
I think for QAT testing we can use pseudo labels generated by the FP32 model, do a few mini training steps on the fake-quant model, and then compare outputs against the FP32 baseline (pseudo labels) within acceptable atol/rtol thresholds [as usual](https://github.com/pytorch/executorch/blob/main/backends/qualcomm/tests/test_qnn_delegate.py#L3497).

sxu commented:
If I manually convert this line https://github.com/pytorch/executorch/blob/d13c78971338b35a219e0a27f7045b29c7f4e8c1/backends/qualcomm/builders/node_visitor.py#L198
to be a Python list it seems to work, perhaps some pybind dependencies weren't being included when running the test?

cccclai reviewed (approved):
Approve on be half on QCOM's review

sxu committed:
Add 16a4w_block QAT config (#15878)

Summary:

Introduce a FakeQuantizer subclass. It falls back to LPBQ observer's `convert`. `_derived_bias_quant_spec` also looks for it to correctly derive bias scale.

Reviewed By: viveknayakatmeta

Differential Revision: D87194388

ares89 commented:
> Please ignore my question above about NumPy array, our internal test runner was missing a pybind dependency.


I'm also experiencing the same issue. May I ask which dependency is missing?
